README.md
## Projekt Systemu Zapisów Instytutu Informatyki Uniwersytetu Wrocławskiego

### Rozpoczynanie pracy

Aby rozpocząć pracę w projekcie przejdź przez instrukcję:
https://github.com/iiuni/projektzapisy/wiki/Developer's-environment-setup

Jeżeli zamierzasz zajmować się rzeczami związanymi z _frontendem_, warto zapoznać się też z systemem plików statycznych:
https://github.com/iiuni/projektzapisy/wiki/Pliki-statyczne-w-Systemie-Zapis%C3%B3w

### Narzędzia

* **Slack** - nasz główny komunikator: https://projektzapisy.slack.com/
* **Redmine** - publiczny tracker błędów: https://tracker-zapisy.ii.uni.wroc.pl/projects/zapisy-tracker/issues
* **Rollbar** - tracker błędów pojawiających się na produkcji: https://rollbar.com/IIUniversityofWroclaw/System-Zapisow/
* **CircleCI** - continuous integration: https://circleci.com/gh/lewymati/projektzapisy

### Workflow pracy

1. Gdy decydujemy się zająć się jakąś funkcjonalnością/bugiem przypisujemy sobie odpowiedni Issue tutaj na Githubie (jeśli go nie ma to, tworzymy go).
2. Pracujemy na swoim branchu - tworzymy go z brancha `master-dev`.
3. Po ukończeniu tworzymy Pull Requesta z bazowym branchem `master-dev`.
4. Wybieramy w PR dwie osoby, do zrobienia Code Review.
5. Po akceptacji dwóch osób możemy samodzielnie kliknąć Merge.
6. Co jakiś czas (około dwóch tygodni-miesiąca) robiony jest deploy na produkcję. Branch `master` przechowuje wersję produkcyjną.

### Wskazówki

* Napotykając na problem warto napisać na Slacku. Jest szansa, że ktoś inny już natrafił na ten problem (a jeśli nie to przynajmniej inne osoby mogą być świadome trudności)
* Starą wersję bazy danych, możesz aktualizować aplikując migracje komendą:
`./manage.py migrate`
* Zmieniając model, możesz zmienić strukturę bazy danych. Poniższą komendą możesz wygenerować automatycznie migrację:
`./manage.py makemigrations`


CHANGELOG.md
# projektzapisy Changelog

# Version 3.6.19

* Link to issue tracker replaced with Github (#680)
* Small interface changes
* Re-write of offer/vote application (#678)
* Re-write of offer/preferences application (#686)

# Version 29.4.19

* New model for course Proposal and base model for Course info (#669)

# Version 25.4.19

* CSS library updated to Bootstrap 4 across entire system (#603)
* CAS server redirection back to the system after log-out (#655)
* Attaching security-related (mainly XSS) headers (#552)
* Upgrade of some Python dependencies, including Django (to v2.1) (#666)

# Version 12.2.19

* New opening times computation (#500).
* New ECTS course values computation (#501).
* New enrollment/queues logic (#502).
* New timetable and prototype (#503).

# Version 31.1.19

* REST API for students votes in offer is introduced to make exporting votes to scheduler easier.
* Bug in key-generation is fixed, that was probably introduced by upgrading Django to 2.0.

# Version 5.12.18

* Ubuntu upgraded to 18.04.
* Major changes in users app. We have groups now. UserProfile and ExtendedUser are gone (#444).
* Webpack/Yarn is used for compiling and bundling assets (JS/TS/CSS).
* Django is updated to version 2.0.

# Version 10.06.18

* added consent for students
* small improvement in desideratas
* special reservations api

# Version 18.05.18

* implemented new backups
* moved rest api to seperate app
* added limit on the number of rest calls made by unauthenticated users (100/day)

# Version 18.04.16

* Upgrade to python3
* Added simple semester REST API serving semesters
* Classroom REST API serving classroom data
* Fixed negative counts in grade summary
* Added confirmation pop-up before direct disenrollment

* Removed redundant informations about isim students
* Removed anything related to program "studia zamawiane"
* Removed UserProfile and ExtendedUser
* Changed authentication backend to django's default

# Version 18.02.25

* Added webpack
* Fixed reports generation
* Changed email signature for news

# Version 18.02.02

* Upgrade to Django 1.11
* Fix email sending

# Version 18.01.24

* Fix changing email
* added news author in emails
* Fixed events' start time in iCalc
* Fixed remove student from queue when enrolling to lecture
* Fixed issue with contextprocessors
* Fixed editing course in offer

# Version 18.01.05

* Upgrade to Django 1.10
* Fix for pulling student from queue

# Version 17.12.16

* Improved desideratas form
* Upgrade to Django 1.9

# Version 17.12.11

* Show desiderates in offer menu
* Upgraded the google analytics library

# Version 17.11.28

* Fixed users schedules
* Fixed course proposal form
* Fixed removing student from group in admin
* Removed paranoid sessions
* Fixed ignoring conflicts on reservation edition
* Cleaned main repository directory

# Version 17.11.21

* Upgrade to Django 1.8
* Loading courses with ajax
* Fixed disappearing pop-up
* Fixed issues with reservations and conflicts
* Hid vote summary from non-users

# Version 17.10.25

* Fixed missing title for login page
* Hid vote button for employees
* Added notification when leaving voting page
* Fixed several typos
* Fixed 'None' in news search field
* Improved login UX
* Added displaying links to courses on employee's page
* Added showing records ending time
* Added USOS CAS authentication

# Version 17.06.19

* Upgrade to Django 1.7
* Added maintenance mode switcher
* Fixed course names in vote results
* Enhanced filtering for offer

# Version 17.06.13

* Added possibility to remove reservations
* Added link to tracker to 500 error page

# Version 17.06.07

* Fixed error with not visible plans and not working prototype when new hidden future semester is added
* Fixed 500 error on vote summary
* Fixed password reset functionality after upgrade to Django 1.6. The issue was with wrong url names

# Version 17.06.06

* Added tag/effects filtering
* Bug-fixes after Django 1.6

# Version 17.05.29

* Upgrade to Django 1.6

# Version 17.05.05

* Improved syllabus form

# Version 17.04.20

* Upgrade to Django 1.5

# Version 17.04.20

* Fixed bug with sending emails with "ó"
* Added possibility to edit syllabus
* Added conflict list for secretary

# Version 17.04.07

* Fixed mail-to links escape '+'
* Fixed proper author while creating cyclical reservation
* Added possibility to add reservations with conflicts for secretary
* Fixed FreeDay and ChangedDay are now unique
* Added news content is now included in emails

# Version 17.03.09

* Added new date for records: end of unenrolling

# Version 17.02.26

* Added rooms reports
* Fixed polish fonts in pdfs
* Fixed export to ical
* Fixed display enrollment number (with isim students)

# Version 17.02.13

* Cyclic reservations can now be added with conflicts
* Fixed tight time bounds for reservations

# Version 17.01.31

* Fixed cyclical reservation starts no earlier than semester beginning
* Fixed open only SMTP one connection while sending emails

# Version 17.01.25

* Updated max points for vote information
* Disabled voting for crossed out students
* Adding new event doesn't require description and event is visible for all users by default
* Added python-rq
* Added some schedule/enrollment/vote tests

# Version 16.12.19

* Added cookie law info
* Added tracker for bugs at tracker-zapisy.ii.uni.wroc.pl
* Changed email from zapisy@ii.uni.wroc.pl to zapisy@cs.uni.wroc.pl
* Added feature to email all students
* Fix for bug with unrolling from last lecture

# Version 16.12.03

* Courses in review are no longer visible at all
* Courses not in offer are no longer visible for not authenticated users
* Courses descriptions are now sorted by time of creation
* Fixed proper Freedays and ChangedDays validation while making reservation

# Version 16.11.18

* Reworked sending emails to group and queues
* Voting link is now not visible when voting is closed + added tests

# Version 16.11.04

* Fixed an issue with dev env
* Added working tests to CI

zapisy/settings.py
# Django settings for fereol project.

import os
import logging

PROJECT_PATH = os.path.abspath(os.path.dirname(__file__))

DEBUG = False

RELEASE = False

TEMPLATE_DEBUG = False

ADMINS = (
    # ('Your Name', 'your_email@domain.com'),
)

MANAGERS = ADMINS

# With DEBUG = False Django will refuse to serve requests to hosts different
# than this one.
ALLOWED_HOSTS = ['zapisy.ii.uni.wroc.pl', 'localhost']
EVENT_MODERATOR_EMAIL = 'zapisy@cs.uni.wroc.pl'

"""
DATABASES = {
     'default' : {
        'ENGINE' : 'postgresql_psycopg2',
        'NAME' : 'fereol',
        'PORT' : '',
        'USER' : 'fereol',
        'PASSWORD' : 'fereol',
        'HOST' : '',
        'CHARSET' : 'utf8',
        'USE_UNICODE' : True,
        },

     'fereol2012' : {
        'ENGINE' : 'postgresql_psycopg2',
        'NAME' : 'fereol2012',
        'PORT' : '',
        'USER' : 'fereol',
        'PASSWORD' : 'fereol',
        'HOST' : '',
        'CHARSET' : 'utf8',
        'USE_UNICODE' : True,
        }
 }
"""


DATABASES = {
     'default' : {
        'ENGINE' : 'django.db.backends.sqlite3',
        'NAME' : os.path.join(PROJECT_PATH, 'database/db.sqlite3'),
        'PORT' : '',
        'USER' : '',
        'PASSWORD' : '',
        'HOST' : '',
        'CHARSET' : 'utf8',
        'USE_UNICODE' : True,
        }
}


# mass-mail account
# You can test sending with:
# $ python -m smtpd -n -c DebuggingServer localhost:1025

# For gmail:
#EMAIL_USE_TLS = True
#EMAIL_HOST = 'smtp.gmail.com'
#EMAIL_HOST_USER = 'youremail@gmail.com'
#EMAIL_HOST_PASSWORD = 'password'
#EMAIL_PORT = 587

#EMAIL_HOST = 'localhost'
#EMAIL_PORT = 1025
#EMAIL_HOST_USER = ''
#EMAIL_HOST_PASSWORD = ''

MASS_MAIL_FROM = 'zapisy@cs.uni.wroc.pl'

EMAIL_COURSE_PREFIX = '[System Zapisow] ' # please don't remove the trailing space

#loggin settings:

#LOG_FILE = os.path.join(PROJECT_PATH, "logs/log.log")
#LOG_LEVEL = logging.NOTSET
#INTERNAL_IPS = ('127.0.0.1',)
#logging.basicConfig(level=LOG_LEVEL, filename=LOG_FILE, format = '%(asctime)s | %(levelname)s | %(message)s')
LOGGING = {
        'version': 1,
        'disable_existing_loggers': False, # keep Django's default loggers
        'formatters': {
                'timestampthread': {
                        'format': "%(asctime)s [%(threadName)-12.12s] [%(levelname)-5.5s] [%(name)-20.20s]  %(message)s",
                },
        },
        'handlers': {
                'logfile': {
                        # optionally raise to INFO to not fill the log file too quickly
                        'level': 'DEBUG', # DEBUG or higher goes to the log file
                        'class':'logging.handlers.RotatingFileHandler',
                        # IMPORTANT: replace with your desired logfile name!
                        'filename': 'logs/djangoproject.log',
                        'maxBytes': 50 * 10**6, # will 50 MB do?
                        'backupCount': 3, # keep this many extra historical files
                        'formatter': 'timestampthread'
                },
        },
        'loggers': {
                'django': { # configure all of Django's loggers
                        'handlers': ['logfile'],
                        'level': 'DEBUG', # set to debug to see e.g. database queries
                },
                'apps': {
                        'handlers': ['logfile'],
                        'level': 'DEBUG',
                },
        },
        'root': {
                'handlers': ['logfile'],
                'level': 'DEBUG'
        }
}


# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# If running in a Windows environment this must be set to the same as your
# system time zone.
TIME_ZONE = 'Europe/Warsaw'

# Language code for this installation. All choices can be found here:
# http://www.i18nguy.com/unicode/language-identifiers.html
LANGUAGE_CODE = 'pl-pl'

# Available languages for using the service. The first one is the default.
ugettext = lambda s: s
LANGUAGES = (
    ('pl', ugettext('Polish')),
    ('en', ugettext('English')),
)

SITE_ID = 1

# If you set this to False, Django will make some optimizations so as not
# to load the internationalization machinery.
USE_I18N = True

# Absolute path to the directory that holds media.
# Example: "/home/media/media.lawrence.com/"
MEDIA_ROOT = 'site_media'

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash if there is a path component (optional in other cases).
# Examples: "http://media.lawrence.com", "http://example.com/media/"
MEDIA_URL = '/site_media/'

USE_ETAGS = True

# Make this unique, and don't share it with anybody.
SECRET_KEY = '6$u2ggeh-!^hxep3s4h$3z&2-+3c@sy7-sy8349+l-1m)9r0fn'

# List of callables that know how to import templates from various sources.

TEMPLATE_LOADERS = (
    ('django.template.loaders.cached.Loader', (

    'django.template.loaders.filesystem.Loader',
    'django.template.loaders.app_directories.Loader',
    'django.template.loaders.eggs.Loader',
     )),
)

#TEMPLATE_LOADERS = (
#    'django.template.loaders.filesystem.Loader',
#    'django.template.loaders.app_directories.Loader',
##    'django.template.loaders.eggs.load_template_source',
#)

TEMPLATE_CONTEXT_PROCESSORS = (
    'django.contrib.messages.context_processors.messages',
    'django.contrib.auth.context_processors.auth',
    "django.core.context_processors.request",
)


# Be careful with the order! I'm aware that SessionMiddleware
# and Authentication both must come before LocalePref which
# must precede LocaleMiddleware, and Common must go afterwards.
MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'middleware.localePrefMiddleware.LocalePrefMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.transaction.TransactionMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    #'debug_toolbar.middleware.DebugToolbarMiddleware',
    #'middleware.mobile_detector.mobileDetectionMiddleware',
    #'middleware.mobileMiddleware.SubdomainMiddleware',
    'middleware.error_handling.ErrorHandlerMiddleware',
    'pipeline.middleware.MinifyHTMLMiddleware',
    'rollbar.contrib.django.middleware.RollbarNotifierMiddleware',
)

ROOT_URLCONF = 'urls'

TEMPLATE_DIRS = (
    os.path.join(PROJECT_PATH, 'templates'),
)

INSTALLED_APPS = (
    'modeltranslation', # needs to be before django.contrib.admin
    # needed from 1.7 onwards to prevent Django from trying to apply
    # migrations when testing (slows down DB setup _a lot_)
    'test_without_migrations',
    
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.sites',
    'django.contrib.admin',
    'django.contrib.admindocs',

    'apps.users',

    'django.contrib.messages',
    'django.contrib.staticfiles',
    'mailer',
    #'south',
    'pipeline',
    'apps.enrollment.courses',
    'apps.enrollment.records',
    'apps.statistics',
    'apps.news',
    'apps.offer.preferences',
    'apps.offer.proposal',
    'apps.offer.vote',
    'apps.offer.desiderata',

    'apps.utils',
    'apps.schedule',
    #'debug_toolbar',
    'apps.grade.poll',
    'apps.grade.ticket_create',
    #'apps.mobile',
    'apps.email_change',
    'django_extensions',
    'django_filters',
    'autoslug',
    'endless_pagination',
    'apps.notifications',
    'django_cas_ng',

    'test_app'
)

MODELTRANSLATION_FALLBACK_LANGUAGES = ('pl',)

AUTHENTICATION_BACKENDS = (
    'apps.users.auth_backend.BetterBackend',
    'django_cas_ng.backends.CASBackend',
)

CAS_SERVER_URL = 'https://login.uni.wroc.pl/cas/'
CAS_CREATE_USER = False
CAS_LOGIN_MSG = u'Sukces! Zalogowano przez USOS (login: %s).'

LOGIN_URL = '/users/login/'
LOGIN_REDIRECT_URL = '/users/'

TEST_RUNNER = 'django.test.runner.DiscoverRunner'

#settings for enrollment
ECTS_BONUS = 5 # ECTS_BONUS * ECTS = abs(t0-t1); set to 7, if changed, change also get_t0_interval()
ECTS_LIMIT = 35
ECTS_FINAL_LIMIT = 45

VOTE_LIMIT = 60

M_PROGRAM = 1
LETURE_TYPE = '1'
QUEUE_PRIORITY_LIMIT = 5

# that's only the example of settings_local.py file contents:
#SESSION_COOKIE_DOMAIN = '.localhost.localhost' # without port number!
#SESSION_COOKIE_DOMAIN = '.localhost.localhost' # without port number!
SESSION_COOKIE_PATH = '/;HttpOnly'
#DEBUG = False
#TEMPLATE_DEBUG = DEBUG

SESSION_EXPIRE_AT_BROWSER_CLOSE = True
# Since Django 1.6 the default session serializer is json, which
# doesn't have as many features, in particular it cannot serialize
# custom objects, and we need this behavior.
SESSION_SERIALIZER = 'django.contrib.sessions.serializers.PickleSerializer'

DEBUG_TOOLBAR_ALLOWED_USERS = [
    "209067", # Tomasz Wasilczyk
    "209138", # Arkadiusz Flinik
    "208934",
    "gosia", "stanislaw"
]

def show_toolbar(request):
    if request.user and request.user.username in DEBUG_TOOLBAR_ALLOWED_USERS:
        return True
    return False


DEBUG_TOOLBAR_CONFIG = {
    'SHOW_TOOLBAR_CALLBACK': show_toolbar,
    'INTERCEPT_REDIRECTS' : False,
}


CACHES = {
    'default': {
        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
        'LOCATION': '127.0.0.1:11211',
        'TIMEOUT': 86400,
        'OPTIONS': {
            'MAX_ENTRIES': 1000
        }
    }
}
        
NEWS_PER_PAGE = 15

# The URL to the issue tracker where users
# can submit issues or bug reports. Used in several templates.
ISSUE_TRACKER_URL = "https://tracker-zapisy.ii.uni.wroc.pl"
# As above, but takes the user straight to the "create new issue" page
ISSUE_TRACKER_NEW_ISSUE_URL = "https://tracker-zapisy.ii.uni.wroc.pl/projects/zapisy-tracker/issues/new"

if os.path.isfile(os.path.join(PROJECT_PATH, 'pipeline.py')):
    execfile(os.path.join(PROJECT_PATH, 'pipeline.py'))

PIPELINE = True
PIPELINE_AUTO = False
PIPELINE_VERSION = True
PIPELINE_YUI_BINARY = 'java -jar libs/yuicompressor-2.4.7.jar'
#PIPELINE_JS_COMPRESSOR = 'pipeline.compressors.jsmin.JSMinCompressor'
#PIPELINE_CSS_COMPRESSOR = 'pipeline.compressors.csstidy.CSSTidyCompressor'

STATIC_URL = '/static/'
STATIC_ROOT =  os.path.join(PROJECT_PATH, 'site_media')
STATICFILES_STORAGE = 'pipeline.storage.PipelineStorage'
PIPELINE_STORAGE = 'pipeline.storage.PipelineFinderStorage'
PIPELINE_VERSIONING = 'pipeline.versioning.hash.MD5Versioning'
STATICFILES_FINDERS = (
  'pipeline.finders.PipelineFinder',
  'django.contrib.staticfiles.finders.FileSystemFinder',
  'django.contrib.staticfiles.finders.AppDirectoriesFinder'
)

local_settings_file = os.path.join(PROJECT_PATH, 'settings_local.py')
if os.path.isfile(local_settings_file):
    execfile(local_settings_file)
